PumpingCfg/UnitElimination.lean

```diff
@@ -115,155 +110,153 @@ lemma generators_prod_diag_subset : g.generators_prod_diag ⊆ g.generators ×ˢ
+def collect_unitPair (nᵢ nₒ : g.NT) (p : g.NT × g.NT) (S : Finset (g.NT × g.NT)) :=
+  if nₒ = p.1 then insert (nᵢ, p.2) S else S
 
-def collect_unitPairs (r : ContextFreeRule T g.NT) (pairs : List (g.NT × g.NT)) :=
+def collect_unitPairs (r : ContextFreeRule T g.NT) (S : List (g.NT × g.NT)) :=
   match r.output with
-  | [Symbol.nonterminal v] => pairs.foldr (collect_unitPair r.input v) {}
+  | [Symbol.nonterminal v] => S.foldr (collect_unitPair r.input v) {}
   | _ => {}
 
-lemma rec_collect_unitPairs_unitPairs {input output : g.NT} {p : g.NT × g.NT}
-    {pairs : List (g.NT × g.NT)} {collection : Finset (g.NT × g.NT)}
-    (hmem : p ∈ pairs.foldr (collect_unitPair input output) collection) :
-    p ∈ collection ∨ ∃ v, (output, v) ∈ pairs ∧ p = (input, v) := by
-  revert collection
-  induction pairs with
+lemma rec_collect_unitPairs_unitPairs {nᵢ nₒ : g.NT} {p : g.NT × g.NT}
+    {L : List (g.NT × g.NT)} {S : Finset (g.NT × g.NT)}
+    (hp : p ∈ L.foldr (collect_unitPair nᵢ nₒ) S) :
+    p ∈ S ∨ ∃ v, (nₒ, v) ∈ L ∧ p = (nᵢ, v) := by
+  revert S
+  induction L with
   | nil =>
-    intro collection h
+    intro S hp
     left
-    exact h
-  | cons hd tl ih =>
-    intro collection
+    exact hp
+  | cons d l ih =>
+    intro S
     unfold collect_unitPair
     simp only [List.foldr_cons, List.mem_cons]
-    split <;> (rename_i heq; intro h)
-    · simp only [Finset.mem_insert] at h
-      cases h with
-      | inl h =>
+    split <;> (rename_i hd; intro hp)
+    · simp only [Finset.mem_insert] at hp
+      cases hp with
+      | inl hpd =>
         right
-        use hd.2
+        use d.2
         constructor
         · left
-          rw [heq]
-        · rw[h]
-      | inr h =>
-        specialize ih h
+          rw [hd]
+        · exact hpd
+      | inr hpS =>
+        specialize ih hpS
         cases ih with
-        | inl h' => left; assumption
-        | inr h' =>
-          obtain ⟨v, hv1, hv2⟩ := h'
+        | inl => left; assumption
+        | inr hlp =>
+          obtain ⟨v, hvl, hpv⟩ := hlp
           right
           use v
-          rw [hv2]
-          exact ⟨Or.inr hv1, rfl⟩
-    · specialize ih h
+          rw [hpv]
+          exact ⟨Or.inr hvl, rfl⟩
+    · specialize ih hp
       cases ih with
-      | inl h' => left; assumption
-      | inr h' =>
-        obtain ⟨v, hv1, hv2⟩ := h'
+      | inl => left; assumption
+      | inr hlp =>
+        obtain ⟨v, hvl, hpv⟩ := hlp
         right
         use v
-        rw [hv2]
-        exact ⟨Or.inr hv1, rfl⟩
-
-lemma collect_unitPairs_unitPair {r : ContextFreeRule T g.NT} (pairs : List (g.NT × g.NT))
-    (hmem : r ∈ g.rules) (hp : ∀ p ∈ pairs, UnitPair p.1 p.2) :
-    ∀ p ∈ collect_unitPairs r pairs, UnitPair p.1 p.2 := by
-  intro p h'
-  unfold collect_unitPairs at h'
-  match heq : r.output with
+        rw [hpv]
+        exact ⟨Or.inr hvl, rfl⟩
+
+lemma collect_unitPairs_unitPair {r : ContextFreeRule T g.NT} (S : List (g.NT × g.NT))
```

Comment:
Why do you call Lists and Finsets L/S

Reply:
And should I also do this at EpsilonElimination

Reply:
Well, I'm kinda wannabe mathematicians, so I guess I just like naming data variables by single letters.

Reply:
Lowercase is probably better tho.